### ISSUE TITLE
Add Support for transposed XYAxisSeries

### ISSUE THREAD

**Issue by Jonarw** [oxyplot/oxyplot#1334]
Related to #46 .

I recently hacked together a `HorizontalBoxplotSeries` to make plots like this possible:
![histogram with boxplot](https://user-images.githubusercontent.com/18097859/62683793-9cad3580-b9bf-11e9-9321-20939925955a.png)

I think it would be really nice to have this in the main library and I'd be willing to make that contribution.

I haven't quite yet decided what would be the best way of doing this. I see three possibilities:

1. Make `HorizontalBoxPlotSeries` inherit from `BoxPlotSeries` and override the relevant stuff (which is a lot: rendering, tracker, updatemaxmin etc.)
That's the quick and ugly way I'm using until now. We probably want to do it a bit more elegant.

2. Make `HorizontalBoxPlotSeries` completely independent from BoxPlotSeries
We could reuse the existing `BoxPlotItem`, but there would be a lot of duplicated code between `BoxPlotSeries` and `HorizontalBoxPlotSeries`.

3. Introduce abstract `BoxPlotSeriesBase` and make both `BoxPlotSeries` and `HorizontalBoxPlotSeries` inherit from it
This is the solution I'm leaning towards at the moment. This would minimize code duplication.
Possibly we would want to rename the properties of `BoxPlotItem` to more generic terms (e.g. `BoxMaximum` instead of `BoxTop`) while keeping the old properties as obsolete.

What do you think? Is there a better solution I haven't thought of?

**Comment by VisualMelon** [comment]
I think an important start would be to make `BoxPlotSeries` support being horizontal (i.e. supports being on a verticle X-axis). Many series support this implicitly by using the axes transformation methods. I've not looked at box-plot, but I can't think why this shouldn't be possible in that case. I'll have a closer look over the weekend. That's seems a needless deficiency in the library.

A more fancy API for a horizontal box plot could perhaps then build on this. Would you be willing to share your code in its current form?

**Comment by Jonarw** [comment]
Huh you are right, that would probably even be enough.
When I initially tried to use the `BoxPlotSeries` with a vertical X axis and it didn't work I kind of assumed there must be a major reason for this. But glancing over the code I can't find any.
I will look into it some more.

**Comment by Jonarw** [comment]
I implemented support for vertical X Axis into the BoxPlotSeries. You can check it out here:
https://github.com/Jonarw/oxyplot/tree/horizontalboxplot

There is still some clean-up and testing to do, but it works fine as proof-of concept.

Some comments:

- Some of the other Series that suport vertical X axes (`RectangleSeries`, `HeatmapSeries`, `HistogramSeries`) all have their own duplicated implementation of the `Orientate` and `GetClippingRect` function. Should this functionality be in the `XYAxisSeries` class? In the above linked branch you can see how this could be done. I have also added the functions `GetOrientatedClippingRect`, `TransformAndOrientate` and `OrientateAndInverseTransform` to `XYAxisSeries` to reduce code duplication and make it easier for derived classes to support vertical X Axes.

- I have added transposed examples for `BoxPlotSeries`, `HistogramSeries` and `RectangleSeries` to make sure they still work as intended

- The RectangleSeries currently does not seem to work correctly with vertical X Axis

What do you guys think of this? 
If there are no objections I will finalize the implementation and look into why the RectangleSeries seems broken when transposed.

**Comment by VisualMelon** [comment]
This all sounds pretty sensible to me. I'll try to look it over properly tomorrow.

**Comment by VisualMelon** [comment]
Comments on the code having actually looked at it:

 - Changes to XYSeries look sensible.

 - Would be nice to add a transposed `HeatMap` with a logarithmic axis using the `HeatMapRenderMethod.Rectangles` render mode as well, since it's a separate implementation with its own complexities beyond `RectangleSeries`. ([Drop in replacement gist](https://gist.github.com/VisualMelon/107a0941d336a0f7ed27af3b2b45f4c6))

 - I'm not wild about the implementation of the horizontal histogram and rectangle series examples: it would be better I think to add a `Transpose` parameter to examples, so that the 'top-level' methods are as bland as possible when they share the same underlying model logic (makes the dependency totally plain, so nobody can accidentally break anything in future).

 - `RectangleSeries` isn't working because it shadows `Transform`, and is using an 'un-orientated' clipping rect. Fix is to use the new `TransformAndOrient`, remove the manual orientation, and use `GetOrientatedClippingRect`.

**Comment by VisualMelon** [comment]
Other series that look trivially patched include:

 - `ScatterSeries<T>`
 - `LineSeries` and derivatives:
   - `StairStepSeries`
   - `TwoColorLineSeries`
   - `ThreeColorLineSeries`
   - `FunctionSeries` (we get this for free)
   - `AreaSeries`
 - `StemSeries`
 - `ContourSeries`

It looks like swapping out `Transform` for `TransformAndOrientate` and `GetClippingRect` for `GetOrientatedClippingRect` would be all there is to do for each of these.

<hr />

Should it maybe be `InverseTransformAndOrientate` rather than `OrientateAndInverseTransform`? There isn't much of a different, but this would have the same feel as `TransformAndOrientate`.

**Comment by Jonarw** [comment]
> I'm not wild about the implementation of the horizontal histogram and rectangle series examples: it would be better I think to add a Transpose parameter to examples, so that the 'top-level' methods are as bland as possible when they share the same underlying model logic (makes the dependency totally plain, so nobody can accidentally break anything in future).

Agree, will change.

> Should it maybe be `InverseTransformAndOrientate` rather than `OrientateAndInverseTransform`? There isn't much of a different, but this would have the same feel as `TransformAndOrientate`.

The wording was kind of intentional: `TransformAndOrientate` first transforms, then orientates, while `OrientateAndInverseTransform` first orientates and then inverse transforms - but I guess it doesn't matter too much in the end.

I'm not completely happy with these functions myself to be honest - people might get confused as to what is the difference between the 'normal' transformation functions and the variants with orientation. I had another thought - how about we'd do it like this:

1. Mark `Transform`, `InverseTransform` and `GetClippingRect` in `XYAxisSeries` as `virtual`
2. Introduce abstract class `TransposableXYAxisSeries` inheriting from `XYAxisSeries`
3. Override `Transform`, `InverseTransform` and `GetClippingRect` in `TransposableXYAxisSeries` using the orientated variants

Then we could make all the relevant series (`HistogramSeries`, `BoxPlotSeries`, `RectangleSeries`, `HeatmapSeries` plus the ones you mentioned) inherit from `TransposableXYAxisSeries` - and they should just work with minimal changes.

The more I think about this the more I like it. I think it makes it easier to understand what you need to do to make a `Series` transposable (i.e. in most cases: just inherit from `TransposableXYAxisSeries` instead of `XYAxisSeries`), reduces the possibilties to make mistakes and avoids cluttering `XYAxisSeries`. What do you think?

**Comment by VisualMelon** [comment]
(Sorry this is so long...)

Making `Transform` virtual would add overhead, but I don't know how significant that would be. A big benefit would be the lack of an API change. A variation would be for `TransposableXYAxisSeries` to simply shadow the methods. This would be essentially the same solution as is already used on a per-series basis, without changing the behaviour of code that handles `XYAxisSeries` generically.

Then again, maybe it would make sense to simply drop the non-orientating behaviour altogether? Series that _can_ support it will then work, and series which don't support it still won't (they will just fail in different ways if they are misused and detect it). This would be a breaking change, but would simplify the API, and mean that old code that was designed for currently non-transposable series which become transposable. These methods are `protected` as well, so the scope for breakage will mostly be limited to non-generic code. I thought long about this yesterday and decided it was a terrible idea, but now I can't think why...

As mentioned, the `Orientate` method (or equivalent) would need to remain so that axes can 'disorientate' when they need to.

On reflection, I think I think some way of overriding/shadowing the `Transform` methods would be necessary (as oppose to providing separate `AndOrientate` methods) to avoid stealthy breaking changes.

<hr />

Note that changes will still need to be made to series which directly access the `Axis.Transform` methods, so whatever is done won't save some series being manually updated, but it won't be a large job and I'd be happy to do that.

<hr />

To summaries this far too long post, I see 3 reasonable options:

 1. Make `Transform` and friends abstract, override in `TransposableXYAxisSeries`
 2. Shadow `Transform` and friends in `TransposableXYAxisSeries`
 3. Make `Transform` and friends orientate by default

I'm struggling to think why I rejected option 3 yesterday... maybe you can explain why it would be a bad idea, because it seems the best option to me at the moment.

**Comment by Jonarw** [comment]
I don't think the overhead of making `Transform` virtual is significant enough for us to worry about it. [[1]](https://stackoverflow.com/questions/530799/what-are-the-performance-implications-of-marking-methods-properties-as-virtual)

Personally I try to avoid shadowing, as it breaks polymorphism and can give unexpected results when using reflection. So in my opinion, good options would be 1.) and 3.).

At the moment I can't see any major downsides to 3.), except that it may be more likely to break stuff in other people's derived classes.

I think 1.) would make sense if there were some `XYAxisSeries` implementations that are NOT transposable by design, i.e. displaying them transposed would make no sense for some reason. In this case, deriving from `TransposableXYAxisSeries` would make it explicit that the `Series` actually is transposable. 
As far as I'm aware, there is no such series implementation in the oxyplot library and I can't think of any case where an `XYAxisSeries` would not be transposable, so maybe option 3.) would be preferable? Unless you remember why this is a terrible idea :wink:

**Comment by VisualMelon** [comment]
The explicitly non-transposable series that spring to mind are `ColumnSeries` and `BarSeries`, which are explicitly complements of each other. However, they are so explicitly non-transposable, that they go to the effort of detecting and throwing if you mis-use them, so it's not a problem if we 'break' them.

**Comment by Jonarw** [comment]
I am actually not sure what the exact difference between `ColumnSeries` and `BarSeries` is.
They don't actually seem to depend on the X Axis being horizontal (at least they don't enforce this), so I wonder, would a transposed `ColumnSeries` look exactly like a `BarSeries`? Anyway this shouldn't really matter for us at this point.

Given that we couldn't break anything that is not already broken by implementing 3.) I think I'm now actually in favor of this option. Unless of course we find a compelling argument against it.

Maybe we could also get a comment from @objorke?

**Comment by VisualMelon** [comment]
I would agree with your assessment, and would at this time support 3. I'll take a closer look at Bar and Column at some point; I think the different is to do with the rendering of labels, but that's just from memory.

**Comment by Jonarw** [comment]
So this kind of increased in scope from what I had in mind initially... But I went ahead with the solution 3) as discussed above and I'm fairly pleased with the result. https://github.com/Jonarw/oxyplot/tree/transposablexyseries

Series that took minimal or sometimes even no effort to patch and are now transposable in my branch:

- `AreaSeries`
- `BoxPlotSeries`
- `ContourSeries`
- `FunctionSeries`
- `HeatMapSeries`
- `HighLowSeries`
- `HistogramSeries`
- `LineSeries`
- `RectangleSeries`
- `ScatterErrorSeries`
- `ScatterSeries`
- `StairStepSeries`
- `StemSeries`
- `ThreeColorLineSeries`
- `TwoColorAreaSeries`
- `TwoColorLineSeries`

I added transposed examples for all of these series and it's all looking good to me.

The following series could in principle also support transposed mode, but are non-trivial to patch:

- `BarSeries` (and derivatives)
- `ColumnSeries` (and derivatives)
- `CandleStickSeries`
- `VolumeSeries`
- `CandleStickAndVolumeSeries`
- `OldCandleStickSeries`

These still work completely normal when used non-transposed. I have added checks to all of these to throw an exception if you try to use them transposed.

It is still not completely clear to me whether a transposed `BarSeries` could completely replace a `ColumnSeries`. I gave it a quick try and the only thing that seemed not to work out-of-the-box were to be the labels, which would be fixable. However I think this is out of the scope of this topic, so I didn't touch these for now.

I will do some more testing and put together a pull request during the next days. Unless of course someone finds an issue with my changes I hadn't thought of until now.

**Comment by VisualMelon** [comment]
I've gone through an checked all the transposed/not-transposed example pairs, and they mostly look fine to me. Couple of comments:

 - `ContourSeries` might want some adjustment with respect to label placement: it tries to place them so that they are vaguely vertical I believe, which means they end up about 90 degrees out when transposed.
 - A transposed `HeatMapSeries` using the rectangles render mode (with labels) would be good, because it is a completely unique code-path. Similarly, a `RectangleSeries` with labels would be good (indeed, any plot which supports labels should probably have them in the transpose, since they seem to be the bit that goes wrong).
 - I'm not sure that `HighLow` makes sense: it's certainly misleading, but honestly the type isn't entirely clear even when it isn't transposed. It would be nice to patch `CandleStickSeries` which presents similar information.

In the code, I think the changes to `XYSeries` are good.

 - The `errorBarVectorX/Y` code in `ScatterErrorSeries` might be better computed with `Orientate` to reduce redundancy.

**Comment by Jonarw** [comment]
> ContourSeries might want some adjustment with respect to label placement: it tries to place them so that they are vaguely vertical I believe, which means they end up about 90 degrees out when transposed.

As far as I can tell it just uses the two middle points of the contour, so the label placement only depends on the order the points are specified.

> A transposed HeatMapSeries using the rectangles render mode (with labels) would be good, because it is a completely unique code-path. Similarly, a RectangleSeries with labels would be good (indeed, any plot which supports labels should probably have them in the transpose, since they seem to be the bit that goes wrong).

Good point. I'll add an example for this and look into other series with labels.

> I'm not sure that HighLow makes sense: it's certainly misleading, but honestly the type isn't entirely clear even when it isn't transposed. It would be nice to patch CandleStickSeries which presents similar information.

I just kind of assume these series make sense for finance people ;). I'll have another look at `CandleStickSeries` - there even is a horizontal Candlestick on this [Wikipedia page](https://en.wikipedia.org/wiki/Candlestick_chart), so this seems to be something that is actually used.

> The errorBarVectorX/Y code in ScatterErrorSeries might be better computed with Orientate to reduce redundancy.

Agree. I think we need a separate `Orientate(...)` for `ScreenVector`.

**Comment by Jonarw** [comment]
`CandleStickSeries` and `VolumeSeries` are patched now.

I don't really feel like patching `CandleStickAndVolumeSeries` because this series seems kind of redundant to me - we could get the same plot by using `CandleStickSeries` and `VolumnSeries` on a shared X axis, or am I missing something? 
Maybe `CandleStickAndVolumeSeries` should be marked as `Obsolete`, but that's not part of this issue. In any case, I think I'm fine with leaving `CandleStickAndVolumeSeries` unpatched because of said redundancy.

Another topic: To reduce code redundancy in `ExampleLibrary`, I decided to add some extension methods for transposing and reversing plots, which make it way easier to add transposed/reversed versions of already existing examples.

Can anyone find any additional issues with my changes?

**Comment by VisualMelon** [comment]
I agree that `CandleStickAndVolumeSeries` doesn't make a whole lot of sense, but if it is a sufficiently common use-case then maybe there should be some provision (not necessarily a dedicated series, but some helper which will assemble all the series and axes etc.). I agree it shouldn't be a priority for making transposable.

<hr />

I think the `PlotModelUtilities` is a great idea. I'm not terribly fond of the interface to the `Transpose` and such which mutates and returns the given `PlotModel`, but for the sake of making it trivial to write examples I won't complain. I would suggest bulking out the inline documentation to make it clear that it is modifying the given `PlotModel`. The switches could do with throwing a more helpful exception when if there is an unrecognised `AxisPosition`, e.g. 

    throw new ArgumentOutOfRangeException($"AxisPosition for Axis {axis.Key} is not recognised");

There doesn't seem to be too much consistency in the library concerning what to throw in such cases.

<hr />

The placement of `using ExampleLibrary.Utilities;` is inconsistent in `HistogramSeriesExamples.cs` and `AreaSeriesExamples.cs`.

<hr />

I like the additions to `OxyRect`. Maybe change `point` to `corner` in the documentation?

**Comment by Jonarw** [comment]
Thanks for your input!

>  The switches could do with throwing a more helpful exception when if there is an unrecognised AxisPosition

This is just what I end up with when I use the 'Generate Switch Labels' refactoring in Visual Studio (could also be a ReSharper feature). For sure it would make sense to replace it with something more meaningful. I think most elegant solution would be a dedicated `EnumException<TEnum>` class - but again I think this is outside the scope of this issue, especially as this is handled pretty inconsistently in the library, as you already mentioned.

I agree with your other points.

Btw. I did some profiling to verify that this change doesn't kill performance. I used the `LineSeries (1M points)` example and profiled it once with the old `XYAxisSeries.Transform()` (without the call to `Orientate()`) and once with the new `XYAxisSeries.Transform()`. The results are pretty conclusive:

The `LineSeries.ExtractNextContiguousLineSegment()` (which transforms the 1M points) had 30 ms or 0.79% of render time in both cases - no performance change whatsoever on my system. The `XYAxisSeries.Transform()` and `XYAxisSeries.Orientate()` functions don't even show up in the profiler in Release mode (they do in Debug), meaning they are completely inlined by the Jitter.
